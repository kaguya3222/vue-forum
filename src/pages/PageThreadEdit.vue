<template>
  <div v-if="thread && text" class="col-full push-top">
    <h1>
      Editing <i>{{ thread.title }}</i>
    </h1>

    <thread-editor
      :title="thread.title"
      :text="text"
      @save="save"
      @cancel="cancel"
    />
  </div>
</template>

<script>
import ThreadEditor from "../components/ThreadEditor";
import { mapGetters, mapActions } from "vuex";

export default {
  name: "PageThreadEdit",
  components: { "thread-editor": ThreadEditor },
  props: {
    threadId: {
      required: true,
      type: String
    }
  },
  computed: {
    ...mapGetters({
      threads: "threads/items",
      posts: "posts/items"
    }),
    thread() {
      return this.threads[this.threadId];
    },
    text() {
      const post = this.posts[this.thread.firstPostId];
      return post ? post.text : null;
    }
  },
  methods: {
    ...mapActions("threads", ["fetchThread"]),
    ...mapActions("posts", ["fetchPost"]),
    async save({ title, text }) {
      const threadId = await this.$store.dispatch("threads/updateThread", {
        threadId: this.threadId,
        title,
        text
      });
      await this.$router.push({
        name: "ThreadShow",
        params: { threadId }
      });
    },
    cancel() {
      this.$router.push({
        name: "ThreadShow",
        params: { threadId: this.threadId }
      });
    }
  },
  created() {
    this.fetchThread({ id: this.threadId }).then(thread => {
      this.fetchPost({ id: thread.firstPostId });
    });
  }
};
</script>

<style scoped></style>
