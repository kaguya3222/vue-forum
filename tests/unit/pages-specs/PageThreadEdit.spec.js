import { createLocalVue, shallowMount } from "@vue/test-utils";
import PageThreadEdit from "../../../src/pages/PageThreadEdit";
import ThreadEditor from "../../../src/components/ThreadEditor";
import Vuex from "vuex";
import threadsStore from "@/store/modules/threads/store";
import postsStore from "@/store/modules/posts/store";
import mockedRootStore from "../mocks/mockedRootStore";

const localVue = createLocalVue();
localVue.use(Vuex);

const rootState = mockedRootStore.state;

describe("PageThreadEdit", () => {
  const theadActions = {
    fetchThread: jest
      .fn()
      .mockResolvedValue({ firstPostId: "thread's firstPostId" })
  };
  const postActions = {
    fetchPost: jest.fn()
  };
  const store = new Vuex.Store({
    modules: {
      threads: {
        namespaced: true,
        state: rootState.threads,
        getters: threadsStore.getters,
        actions: theadActions
      },
      posts: {
        namespaced: true,
        state: rootState.posts,
        getters: postsStore.getters,
        actions: postActions
      }
    }
  });
  const $router = {
    push: jest.fn()
  };
  const wrapper = shallowMount(PageThreadEdit, {
    propsData: {
      threadId: "-KsjWehQ--apjDBwSBCY"
    },
    mocks: {
      $router
    },
    localVue,
    store
  });
  store.dispatch = jest.fn();
  test("Calls fetchThread action when created", () => {
    expect(theadActions.fetchThread).toHaveBeenCalled();
  });
  test("Correctly shows thread edition page", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("Accepts ThreadEditor as a child component", () => {
    expect(wrapper.contains(ThreadEditor)).toBe(true);
  });
  describe("save method", () => {
    wrapper.vm.$on("save", wrapper.vm.save);
    beforeEach(() => {
      wrapper.vm.$emit("save", { text: "", title: "" });
    });
    test("Dispatches 'updateThread' when 'save' event is emitted", () => {
      expect(store.dispatch).toHaveBeenCalledWith("threads/updateThread", {
        threadId: wrapper.vm.threadId,
        title: expect.any(String),
        text: expect.any(String)
      });
    });
    test("Redirects user to 'ThreadShow' route after form is submitted", () => {
      expect($router.push).toHaveBeenCalledWith({
        name: "ThreadShow",
        params: expect.any(Object)
      });
    });
  });
  describe("cancel", function() {
    wrapper.vm.$on("cancel", wrapper.vm.cancel);
    test("Redirects user to thread page when 'cancel' event is emitted", () => {
      wrapper.vm.$emit("cancel");
      expect($router.push).toHaveBeenCalledWith({
        name: "ThreadShow",
        params: {
          threadId: wrapper.vm.threadId
        }
      });
    });
  });
});
