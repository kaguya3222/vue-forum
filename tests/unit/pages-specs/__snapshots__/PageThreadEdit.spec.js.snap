// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PageThreadEdit Correctly shows thread edition page 1`] = `
<div class="col-full push-top">
  <h1>
    Editing <i>What kind of pet do you have?</i></h1>
  <thread-editor-stub text="I'm curious to know what kind of pets the forum users have. I thought this could be a nice way to get to know each other a little better, so I'll start!

My dog is a five month old Italian mastiff (Cane Corso). Since I got him from Barcelona I thought a Spanish name would be suitable, and I named him Don Jefe, or Jefe for short (which means &quot;boss man&quot;).

Jefe is a cute, little beast who eats 1.5 kg raw meat every day. He love to fart, snore and play.

Here's Jefe with his Instagram hat and sqeaky ball ðŸŽ‰ðŸŽ‰ðŸŽ‰
![Don Jefe](http://i.imgur.com/XCWzR6l.jpg)
" title="What kind of pet do you have?"></thread-editor-stub>
</div>
`;
